src/graph/serialize/dglgraph_serialize.cc

```diff
@@ -0,0 +1,251 @@
+                              std::vector<dgl_id_t> idx_list, bool onlyMeta) {
+  StorageMetaData metadata = StorageMetaData::Create();
+  // Read Graph MetaData
+  dgl_id_t num_graph;
+  CHECK(fs->Read(&num_graph)) << "Invalid num of graph";
+  std::vector<dgl_id_t> graph_indices;
+  std::vector<int64_t> nodes_num_list;
+  std::vector<int64_t> edges_num_list;
+  std::vector<NamedTensor> labels_list;
+
+  CHECK(fs->Read(&graph_indices)) << "Invalid graph indices";
+  CHECK(fs->Read(&nodes_num_list)) << "Invalid node num list";
+  CHECK(fs->Read(&edges_num_list)) << "Invalid edge num list";
+  CHECK(fs->Read(&labels_list)) << "Invalid label list";
+
+  metadata->SetMetaData(num_graph, nodes_num_list, edges_num_list, labels_list);
+
+  std::vector<GraphData> gdata_refs;
+
+  // Early Return
+  if (onlyMeta) {
+    return metadata;
+  }
+
+  if (idx_list.empty()) {
+    // Read All Graphs
+    gdata_refs.reserve(num_graph);
+    for (uint64_t i = 0; i < num_graph; ++i) {
+      GraphData gdata = GraphData::Create();
+      GraphDataObject *gdata_ptr =
+        const_cast<GraphDataObject *>(gdata.as<GraphDataObject>());
+      fs->Read(gdata_ptr);
+      gdata_refs.push_back(gdata);
+    }
+  } else {
+    // Read Selected Graphss
+    gdata_refs.reserve(idx_list.size());
+    // Would be better if idx_list is sorted. However the returned the graphs
+    // should be the same order as the idx_list
+    for (uint64_t i = 0; i < idx_list.size(); ++i) {
+      fs->Seek(graph_indices[idx_list[i]]);
+      GraphData gdata = GraphData::Create();
+      GraphDataObject *gdata_ptr =
+        const_cast<GraphDataObject *>(gdata.as<GraphDataObject>());
+      fs->Read(gdata_ptr);
+      gdata_refs.push_back(gdata);
+    }
+  }
+  metadata->SetGraphData(gdata_refs);
+  return metadata;
+}
+
+void GraphDataObject::SetData(ImmutableGraphPtr gptr,
+                              Map<std::string, Value> node_tensors,
+                              Map<std::string, Value> edge_tensors) {
+  this->gptr = gptr;
+
+  for (auto kv : node_tensors) {
+    std::string name = kv.first;
+    Value v = kv.second;
+    NDArray ndarray = static_cast<NDArray>(v->data);
+    this->node_tensors.emplace_back(name, ndarray);
+  }
+  for (auto kv : edge_tensors) {
+    std::string &name = kv.first;
+    Value v = kv.second;
+    const NDArray &ndarray = static_cast<NDArray>(v->data);
+    this->edge_tensors.emplace_back(name, ndarray);
+  }
+}
+
+void GraphDataObject::Save(dmlc::Stream *fs) const {
+  // Using in csr for storage
+  const CSRPtr g_csr = this->gptr->GetInCSR();
+  fs->Write(g_csr->indptr());
+  fs->Write(g_csr->indices());
+  fs->Write(g_csr->edge_ids());
+  fs->Write(node_tensors);
+  fs->Write(edge_tensors);
+}
+
+bool GraphDataObject::Load(dmlc::Stream *fs) {
+  NDArray indptr, indices, edge_ids;
+  fs->Read(&indptr);
+  fs->Read(&indices);
+  fs->Read(&edge_ids);
+  this->gptr = ImmutableGraph::CreateFromCSR(indptr, indices, edge_ids, "in");
```

Comment:
If the graph have restricted format, shall we re-build that format during load?
Also save the graph format preference.

Reply:
This one is for the legacy storage for `DGLGraph`, which we decided to store Immutable graph for all `DGLGraph`. For the detailed implementation of `HeteroGraph` you can see from https://github.com/VoVAllen/dgl/blob/c2e61ce10b4704ce24133900731787c0bcc5847e/src/graph/unit_graph.cc#L1458 and https://github.com/VoVAllen/dgl/blob/c2e61ce10b4704ce24133900731787c0bcc5847e/src/graph/heterograph.cc#L396

Reply:
One question would be  whether the save and load implementation on `UnitGraph` and `HeteroGraph` would be a stable(won't change in the near feature) implementation? @jermainewang